cpp/src/arrow/compute/expression.cc

```diff
@@ -845,7 +845,8 @@ Result<Expression> FoldConstants(Expression expr) {
       std::move(expr), [](Expression expr) { return expr; },
       [](Expression expr, ...) -> Result<Expression> {
         auto call = CallNotNull(expr);
-        if (std::all_of(call->arguments.begin(), call->arguments.end(),
+        if (!call->arguments.empty() &&
```

Comment:
There could be *pure* functions called with zero arguments, in which case constant folding would not be incorrect. I think the better fix here is to add a `Function::is_impure()` property (defaults to `false`, overridden to return `true` for random). For impure functions, we should probably skip any simplification of the call itself (but simplifying its arguments should be fine)

Reply:
Thanks for your review. The `is_impure` is a better way to decide whether a call itself need simplified, and we need to consider the functions with no arguments during registered(random, hash_count_all, count_all, udf-functions) . So maybe we can implement `is_impure` property by return `Function::arity().num_args == 0 && Function::arity().is_varargs == false`  rather than overridden for every impure function?.

> There could be _pure_ functions called with zero arguments, in which case constant folding would not be incorrect.

Besides, I don't understand the case that _pure_ functions called with zero arguments. You mean a function expression with varargs and input `call->arguments` is zero, or the `guaranteed_true_predicate` expression's `FoldConstants`.

Reply:
> So maybe we can implement is_impure property by return Function::arity().num_args == 0 && Function::arity().is_varargs == false rather than overridden for every impure function?

I don't think that's viable; for example there may be a UDF which is impure but still takes arguments.

> There could be `pure` functions called with zero arguments

For example, a UDF `third_party_magic_number()` which takes no arguments but always returns a compile time constant is pure since the return value is identical for identical arguments. In this example, we should be able to fold the filter expression `add(third_party_magic_number(), 42) >= 777` into a boolean scalar, potentially skipping a lot of scanning

Reply:
Oh, i see, agree with your suggestion!

Reply:
I have add the is_impure property. But seems override `Function::is_impure()` is not convenient. Because the built-in functions are all added by `HashAggregateFunction`,`ScalarAggregateFunction` and other basic inherited classes.
So i have added an init opts, then we could set the is_impure property when we add the specific function into kernel.
How about this method?

Reply:
Also, it's worth noting that currently only ScalarFunctions appear in Expressions and so they're the only ones where expression simplification might happen

Reply:
Yes, you're right. The init opts for all function's inherited classes is unnecessary. I added the impure init opt only for ScalarFunction, it's reasonable.